nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/test/java/org/apache/nifi/oauth2/StandardOauth2AccessTokenProviderTest.java

```diff
@@ -109,40 +119,251 @@ public void testAcquireNewToken() throws Exception {
+
+            if (callCounter.get() == 1) {
+                return buildSuccessfulInitResponse();
+            } else if (callCounter.get() == 2) {
+                throw new IOException(refreshErrorMessage);
+            } else if (callCounter.get() == 3) {
+                return successfulAcquireResponse;
+            }
+
+            throw new IllegalStateException("Test improperly defined mock HTTP responses.");
+        });
+
+        // Get a good accessDetails so we can have a refresh a second time
+        testSubject.getAccessDetails();
+
+        // WHEN
+        String actualToken = testSubject.getAccessDetails().getAccessToken();
+
+        // THEN
+        checkLoggedDebugWhenRefreshThrowsIOException();
+
+        checkLoggedRefreshError(new UncheckedIOException("OAuth2 access token request failed", new IOException(refreshErrorMessage)));
+
+        assertEquals(expectedToken, actualToken);
+    }
+
+    @Test
+    public void testHTTPErrorDuringRefreshAndSubsequentAcquire() throws Exception {
+        // GIVEN
+        String errorRefreshResponseBody = "{ \"error_response\":\"refresh_error\" }";
+        String errorAcquireResponseBody = "{ \"error_response\":\"acquire_error\" }";
+
+        Response errorRefreshResponse = buildResponse(500, errorRefreshResponseBody);
+        Response errorAcquireResponse = buildResponse(503, errorAcquireResponseBody);
+
+        AtomicInteger callCounter = new AtomicInteger(0);
+        when(mockHttpClient.newCall(any(Request.class)).execute()).thenAnswer(invocation -> {
+            callCounter.incrementAndGet();
+
+            if (callCounter.get() == 1) {
+                return buildSuccessfulInitResponse();
+            } else if (callCounter.get() == 2) {
+                return errorRefreshResponse;
+            } else if (callCounter.get() == 3) {
+                return errorAcquireResponse;
+            }
+
+            throw new IllegalStateException("Test improperly defined mock HTTP responses.");
+        });
+
+        List<String> expectedLoggedInfo = Arrays.asList(
+            String.format("OAuth2 access token request failed [HTTP %d], response:%n%s", 500, errorRefreshResponseBody),
+            String.format("OAuth2 access token request failed [HTTP %d], response:%n%s", 503, errorAcquireResponseBody)
+        );
+
+        // Get a good accessDetails so we can have a refresh a second time
+        testSubject.getAccessDetails();
+
+        // WHEN
+        try {
+            testSubject.getAccessDetails();
+            fail();
+        } catch (ProcessException e) {
+            // THEN
+            checkLoggedDebugWhenRefreshThrowsIOException();
+
+            checkLoggedRefreshError(new ProcessException("OAuth2 access token request failed [HTTP 500]"));
+
+            checkedLoggedErrorWhenRefreshReturnsBadHTTPResponse(expectedLoggedInfo);
+
+            checkError(new ProcessException("OAuth2 access token request failed [HTTP 503]"), e);
+        }
```

Comment:
With recent work underway to refactor unit tests, it would be helpful to implement this test using `assertThrows()`.

Reply:
The Junit5 `assertThrow` support is kind of up to par to the current JavaSE solution, although even that is debatable. It is fairly new and hasn't been proven to be obviously superior yet.

The Junit4 approach however has objective shortcomings.
Switching to Junit5 is out of scope of this change.

Leaving the tests as they are is the most optimal in my opinion.

Reply:
JUnit 4.13 includes `assertThrows()`, which is already used in other places throughout the system. Using `assertThrows` with the current JUnit 4 test methods should be a straightforward adjustment, and will also streamline future migration to JUnit 5.

Reply:
Updated.